pyamg/classical/interpolate.py

```diff
@@ -38,34 +38,208 @@ def direct_interpolation(A, C, splitting):
+        amg_core.mod_standard_interpolation_pass2(A.shape[0], A.indptr, A.indices,
+                                                  A.data, C.indptr, C.indices,
+                                                  C.data, splitting, P_indptr,
+                                                  P_colinds, P_data)
+    else:
+        amg_core.rs_standard_interpolation_pass2(A.shape[0], A.indptr, A.indices,
+                                                 A.data, C.indptr, C.indices,
+                                                 C.data, splitting, P_indptr,
+                                                 P_colinds, P_data)
+    
+    return csr_matrix((P_data, P_colinds, P_indptr), shape=[n,nc])
+
+
+def distance_two_interpolation(A, C, splitting, theta=None, norm='min', plus_i=False):
+    """Create prolongator using distance-two AMG interpolation (extended+i interpolaton).
+
+    Parameters
+    ----------
+    A : {csr_matrix}
+        NxN matrix in CSR format
+    C : {csr_matrix}
+        Strength-of-Connection matrix
+        Must have zero diagonal
+    splitting : array
+        C/F splitting stored in an array of length N
+    theta : float in [0,1), default None
+        theta value defining strong connections in a classical AMG
+        sense. Provide if a different SOC is used for P than for
+        CF-splitting; otherwise, theta = None. 
+    norm : string, default 'abs'
+        Norm used in redefining classical SOC. Options are 'min' and
+        'abs' for CSR matrices. See strength.py for more information.
+    plus_i : bool, default True
+        Use "Extended+i" interpolation from [0] as opposed to "Extended"
+        interpolation. Can give better interpolation with minimal
+        added expense.
+
+    Returns
+    -------
+    P : {csr_matrix}
+        Prolongator using standard interpolation
+
+    References
+    ----------
+    [0] "Distance-Two Interpolation for Parallel Algebraic Multigrid,"
+       H. De Sterck, R. Falgout, J. Nolting, U. M. Yang, (2007).
 
-    amg_core.rs_direct_interpolation_pass1(A.shape[0],
-                                           C.indptr, C.indices, splitting, Pp)
+    Examples
+    --------
+    >>> from pyamg.gallery import poisson
+    >>> from pyamg.classical.interpolate import distance_two_interpolation
+    >>> import numpy as np
+    >>> A = poisson((9,),format='csr')
+    >>> splitting = np.array([1,0,1,0,1], dtype='intc')
+    >>> P = distance_two_interpolation(A, A, splitting, 0.25)
+    >>> print(P.todense())
+    [[ 1.   0.   0. ]
+     [ 0.5  0.5  0. ]
+     [ 0.   1.   0. ]
+     [ 0.   0.5  0.5]
+     [ 0.   0.   1. ]]
 
-    nnz = Pp[-1]
-    Pj = np.empty(nnz, dtype=Pp.dtype)
-    Px = np.empty(nnz, dtype=A.dtype)
+    """
+    if not isspmatrix_csr(C):
+        raise TypeError('Expected csr_matrix SOC matrix, C.')
+
+    nc = np.sum(splitting)
+    n = A.shape[0]
+
+    if theta is not None:
+        C = classical_strength_of_connection(A, theta=theta, norm=norm)
```

Comment:
`else:`
`   C = C.copy()`
we need this copy, or else we'll overwrite the entries in `C`